// ra-core/dist/test-ui/Confirm.cjs
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Confirm = void 0;
var __1 = require("..//index.cjs");
var React = __importStar(require("react"));
var Confirm = function (_a) {
    var isOpen = _a.isOpen, content = _a.content, onClose = _a.onClose, onConfirm = _a.onConfirm, title = _a.title, _b = _a.translateOptions, translateOptions = _b === void 0 ? {} : _b, _c = _a.titleTranslateOptions, titleTranslateOptions = _c === void 0 ? translateOptions : _c, _d = _a.contentTranslateOptions, contentTranslateOptions = _d === void 0 ? translateOptions : _d;
    return isOpen ? (React.createElement("div", { style: {
            position: 'fixed',
            top: 0,
            left: 0,
            right: 0,
            bottom: 0,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: 'rgba(0, 0, 0, 0.3)',
        } },
        React.createElement("div", { style: {
                backgroundColor: 'rgba(0, 0, 0, 0.8)',
                color: 'white',
                padding: '1em',
            } },
            React.createElement("p", null, typeof title === 'string' ? (React.createElement(__1.Translate, { i18nKey: title, options: __assign({ _: title }, titleTranslateOptions) })) : (title)),
            React.createElement("p", null, typeof content === 'string' ? (React.createElement(__1.Translate, { i18nKey: content, options: __assign({ _: content }, contentTranslateOptions) })) : (content)),
            React.createElement("div", { style: { display: 'flex', gap: '1em' } },
                React.createElement("button", { onClick: onConfirm, type: "button" },
                    React.createElement(__1.Translate, { i18nKey: "ra.action.confirm" }, "Confirm")),
                React.createElement("button", { onClick: onClose, type: "button" },
                    React.createElement(__1.Translate, { i18nKey: "ra.action.cancel" }, "Cancel")))))) : null;
};
exports.Confirm = Confirm;
