// ra-core/dist/test-ui/DataTable.cjs
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTable = void 0;
var React = __importStar(require("react"));
var __1 = require("..//index.cjs");
var react_router_1 = require("react-router");
var DataTableCol = function (props) {
    var renderContext = (0, __1.useDataTableRenderContext)();
    switch (renderContext) {
        case 'header':
            return React.createElement(DataTableHeadCell, __assign({}, props));
        case 'data':
            return React.createElement(DataTableCell, __assign({}, props));
    }
};
var DataTableHeadCell = function (props) {
    var _a, _b, _c;
    return (React.createElement("th", null, (_a = props.label) !== null && _a !== void 0 ? _a : (React.createElement(React.Fragment, null, (_b = props.source) === null || _b === void 0 ? void 0 :
        _b.substring(0, 1).toUpperCase(), (_c = props.source) === null || _c === void 0 ? void 0 :
        _c.substring(1)))));
};
var DataTableCell = function (props) {
    var record = (0, __1.useRecordContext)();
    if (props.render) {
        return React.createElement("td", null, props.render(record));
    }
    if (props.children) {
        return React.createElement("td", null, props.children);
    }
    if (props.field) {
        return (React.createElement("td", null, React.createElement(props.field, { source: props.source })));
    }
    if (props.source) {
        return (React.createElement("td", null,
            React.createElement(DataTableCellValue, { source: props.source })));
    }
};
var DataTableCellValue = function (props) {
    var value = (0, __1.useFieldValue)(props);
    return React.createElement(React.Fragment, null, value);
};
var DataTableRow = function (props) {
    var getPathForRecord = (0, __1.useGetPathForRecordCallback)();
    var navigate = (0, react_router_1.useNavigate)();
    var record = (0, __1.useRecordContext)(props);
    if (!record) {
        throw new Error('DataTableRow can only be used within a RecordContext or be passed a record prop');
    }
    var resource = (0, __1.useResourceContext)(props);
    if (!resource) {
        throw new Error('DataTableRow can only be used within a ResourceContext or be passed a resource prop');
    }
    var rowClick = (0, __1.useDataTableCallbacksContext)().rowClick;
    var handleClick = (0, __1.useEvent)(function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var temporaryLink, link, _a, path;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    event.persist();
                    temporaryLink = typeof rowClick === 'function'
                        ? rowClick(record.id, resource, record)
                        : rowClick;
                    if (!isPromise(temporaryLink)) return [3 /*break*/, 2];
                    return [4 /*yield*/, temporaryLink];
                case 1:
                    _a = _b.sent();
                    return [3 /*break*/, 3];
                case 2:
                    _a = temporaryLink;
                    _b.label = 3;
                case 3:
                    link = _a;
                    return [4 /*yield*/, getPathForRecord({
                            record: record,
                            resource: resource,
                            link: link,
                        })];
                case 4:
                    path = _b.sent();
                    if (path === false || path == null) {
                        return [2 /*return*/];
                    }
                    navigate(path, {
                        state: { _scrollToTop: true },
                    });
                    return [2 /*return*/];
            }
        });
    }); });
    return React.createElement("tr", { onClick: handleClick }, props.children);
};
var isPromise = function (value) {
    return value && typeof value.then === 'function';
};
var DataTable = function (props) {
    var data = (0, __1.useListContext)().data;
    return (React.createElement(__1.DataTableBase, __assign({}, props, { hasBulkActions: false, empty: null, loading: null }),
        React.createElement("table", { border: 1, style: { width: '100%', borderCollapse: 'collapse' } },
            React.createElement(__1.DataTableRenderContext.Provider, { value: "header" },
                React.createElement("thead", null,
                    React.createElement("tr", null, props.children))),
            React.createElement(__1.DataTableRenderContext.Provider, { value: "data" },
                React.createElement("tbody", null, data === null || data === void 0 ? void 0 : data.map(function (record) { return (React.createElement(__1.RecordContextProvider, { key: record.id, value: record },
                    React.createElement(DataTableRow, null, props.children))); }))))));
};
exports.DataTable = DataTable;
exports.DataTable.Col = DataTableCol;
