// ra-core/dist/test-ui/Layout.js
import { useRefresh, useResourceDefinitions, useTranslate } from "..//index.js";
import * as React from 'react';
import { Link } from 'react-router-dom';
import { Notification } from "./Notification.js";
export var Layout = function (_a) {
    var children = _a.children;
    var resources = useResourceDefinitions();
    var translate = useTranslate();
    var refresh = useRefresh();
    return (React.createElement("div", { style: { padding: '1em', border: '1px solid #ccc' } },
        React.createElement("header", null,
            React.createElement("nav", { style: {
                    display: 'flex',
                    gap: '1em',
                    listStyle: 'none',
                    marginBlockStart: 0,
                    marginBlockEnd: 0,
                    paddingInlineStart: 0,
                    justifyContent: 'space-between',
                } },
                React.createElement("ul", { style: {
                        display: 'flex',
                        gap: '1em',
                        listStyle: 'none',
                        marginBlockStart: 0,
                        marginBlockEnd: 0,
                        paddingInlineStart: 0,
                    } }, Object.values(resources).map(function (resource) {
                    return resource.hasList ? (React.createElement("li", { key: resource.name },
                        React.createElement(Link, { to: "/".concat(resource.name) }, "".concat(resource.name
                            .substring(0, 1)
                            .toUpperCase()).concat(resource.name.substring(1))))) : null;
                })),
                React.createElement("div", { style: {
                        display: 'flex',
                        gap: '1em',
                        listStyle: 'none',
                        marginBlockStart: 0,
                        marginBlockEnd: 0,
                        paddingInlineStart: 0,
                    } },
                    React.createElement("button", { onClick: function () { return refresh(); } }, translate('ra.action.refresh'))))),
        React.createElement("hr", null),
        React.createElement("main", null, children),
        React.createElement(Notification, null)));
};
