// ra-core/dist/controller/input/ArrayInputBase.cjs
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getArrayInputError = exports.ArrayInputBase = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var react_hook_form_1 = require("react-hook-form");
var validate_1 = require("../../form/validation/validate.cjs");
var useGetValidationErrorMessage_1 = require("../../form/validation/useGetValidationErrorMessage.cjs");
var useApplyInputDefaultValues_1 = require("../../form/useApplyInputDefaultValues.cjs");
var useFormGroupContext_1 = require("../../form/groups/useFormGroupContext.cjs");
var useFormGroups_1 = require("../../form/groups/useFormGroups.cjs");
var core_1 = require("../../core/index.cjs");
var ArrayInputContext_1 = require("./ArrayInputContext.cjs");
/**
 * To edit arrays of data embedded inside a record, <ArrayInputBase> creates a list of sub-forms.
 *
 *  @example
 *
 *      import { ArrayInputBase } from 'ra-core';
 *      import { SimpleFormIterator, DateInput, TextInput } from 'my-react-admin-ui';
 *
 *      <ArrayInputBase source="backlinks">
 *          <SimpleFormIterator>
 *              <DateInput source="date" />
 *              <TextInput source="url" />
 *          </SimpleFormIterator>
 *      </ArrayInputBase>
 *
 * <ArrayInputBase> allows the edition of embedded arrays, like the backlinks field
 * in the following post record:
 *
 * {
 *   id: 123
 *   backlinks: [
 *         {
 *             date: '2012-08-10T00:00:00.000Z',
 *             url: 'http://example.com/foo/bar.html',
 *         },
 *         {
 *             date: '2012-08-14T00:00:00.000Z',
 *             url: 'https://blog.johndoe.com/2012/08/12/foobar.html',
 *         }
 *    ]
 * }
 *
 * <ArrayInputBase> expects a single child, which must be a *form iterator* component.
 * A form iterator is a component accepting a fields object as passed by
 * react-hook-form-arrays's useFieldArray() hook, and defining a layout for
 * an array of fields.
 *
 * @see {@link https://react-hook-form.com/docs/usefieldarray}
 */
var ArrayInputBase = function (props) {
    var children = props.children, _a = props.defaultValue, defaultValue = _a === void 0 ? [] : _a, resourceFromProps = props.resource, arraySource = props.source, validate = props.validate;
    var formGroupName = (0, useFormGroupContext_1.useFormGroupContext)();
    var formGroups = (0, useFormGroups_1.useFormGroups)();
    var parentSourceContext = (0, core_1.useSourceContext)();
    var finalSource = parentSourceContext.getSource(arraySource);
    var sanitizedValidate = Array.isArray(validate)
        ? (0, validate_1.composeSyncValidators)(validate)
        : validate;
    var getValidationErrorMessage = (0, useGetValidationErrorMessage_1.useGetValidationErrorMessage)();
    var getValues = (0, react_hook_form_1.useFormContext)().getValues;
    var fieldProps = (0, react_hook_form_1.useFieldArray)({
        name: finalSource,
        rules: {
            validate: function (value) { return __awaiter(void 0, void 0, void 0, function () {
                var error;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!sanitizedValidate)
                                return [2 /*return*/, true];
                            return [4 /*yield*/, sanitizedValidate(value, getValues(), props)];
                        case 1:
                            error = _a.sent();
                            if (!error)
                                return [2 /*return*/, true];
                            return [2 /*return*/, getValidationErrorMessage(error)];
                    }
                });
            }); },
        },
    });
    (0, react_1.useEffect)(function () {
        if (formGroups && formGroupName != null) {
            formGroups.registerField(finalSource, formGroupName);
        }
        return function () {
            if (formGroups && formGroupName != null) {
                formGroups.unregisterField(finalSource, formGroupName);
            }
        };
    }, [finalSource, formGroups, formGroupName]);
    (0, useApplyInputDefaultValues_1.useApplyInputDefaultValues)({
        inputProps: __assign(__assign({}, props), { defaultValue: defaultValue }),
        isArrayInput: true,
        fieldArrayInputControl: fieldProps,
    });
    // The SourceContext will be read by children of ArrayInput to compute their composed source and label
    //
    // <ArrayInput source="orders" /> => SourceContext is "orders"
    //   <SimpleFormIterator> => SourceContext is "orders.0"
    //     <DateInput source="date" /> => final source for this input will be "orders.0.date"
    //   </SimpleFormIterator>
    // </ArrayInput>
    //
    var sourceContext = React.useMemo(function () { return ({
        // source is the source of the ArrayInput child
        getSource: function (source) {
            if (!source) {
                // SimpleFormIterator calls getSource('') to get the arraySource
                return parentSourceContext.getSource(arraySource);
            }
            // We want to support nesting and composition with other inputs (e.g. TranslatableInputs, ReferenceOneInput, etc),
            // we must also take into account the parent SourceContext
            //
            // <ArrayInput source="orders" /> => SourceContext is "orders"
            //   <SimpleFormIterator> => SourceContext is "orders.0"
            //      <DateInput source="date" /> => final source for this input will be "orders.0.date"
            //      <ArrayInput source="items" /> => SourceContext is "orders.0.items"
            //          <SimpleFormIterator> => SourceContext is "orders.0.items.0"
            //              <TextInput source="reference" /> => final source for this input will be "orders.0.items.0.reference"
            //          </SimpleFormIterator>
            //      </ArrayInput>
            //   </SimpleFormIterator>
            // </ArrayInput>
            return parentSourceContext.getSource("".concat(arraySource, ".").concat(source));
        },
        // if Array source is items, and child source is name, .0.name => resources.orders.fields.items.name
        getLabel: function (source) {
            return parentSourceContext.getLabel("".concat(arraySource, ".").concat(source));
        },
    }); }, [parentSourceContext, arraySource]);
    return (React.createElement(ArrayInputContext_1.ArrayInputContext.Provider, { value: fieldProps },
        React.createElement(core_1.OptionalResourceContextProvider, { value: resourceFromProps },
            React.createElement(core_1.SourceContextProvider, { value: sourceContext }, children))));
};
exports.ArrayInputBase = ArrayInputBase;
var getArrayInputError = function (error) {
    if (Array.isArray(error)) {
        return undefined;
    }
    return error;
};
exports.getArrayInputError = getArrayInputError;
