// ra-core/dist/controller/input/SimpleFormIteratorItemBase.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFormIteratorItemBase = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var core_1 = require("../../core/index.cjs");
var useSimpleFormIterator_1 = require("./useSimpleFormIterator.cjs");
var SimpleFormIteratorItemContext_1 = require("./SimpleFormIteratorItemContext.cjs");
var SimpleFormIteratorItemBase = function (props) {
    var children = props.children, index = props.index;
    var resource = (0, core_1.useResourceContext)(props);
    if (!resource) {
        throw new Error('SimpleFormIteratorItem must be used in a ResourceContextProvider or be passed a resource prop.');
    }
    var _a = (0, useSimpleFormIterator_1.useSimpleFormIterator)(), total = _a.total, reOrder = _a.reOrder, remove = _a.remove;
    var context = (0, react_1.useMemo)(function () { return ({
        index: index,
        total: total,
        reOrder: function (newIndex) { return reOrder(index, newIndex); },
        remove: function () { return remove(index); },
    }); }, [index, total, reOrder, remove]);
    var parentSourceContext = (0, core_1.useSourceContext)();
    var sourceContext = (0, react_1.useMemo)(function () { return ({
        getSource: function (source) {
            if (!source) {
                // source can be empty for scalar values, e.g.
                // <ArrayInput source="tags" /> => SourceContext is "tags"
                //   <SimpleFormIterator> => SourceContext is "tags.0"
                //      <TextInput /> => use its parent's getSource so finalSource = "tags.0"
                //   </SimpleFormIterator>
                // </ArrayInput>
                return parentSourceContext.getSource("".concat(index));
            }
            else {
                // Normal input with source, e.g.
                // <ArrayInput source="orders" /> => SourceContext is "orders"
                //   <SimpleFormIterator> => SourceContext is "orders.0"
                //      <DateInput source="date" /> => use its parent's getSource so finalSource = "orders.0.date"
                //   </SimpleFormIterator>
                // </ArrayInput>
                return parentSourceContext.getSource("".concat(index, ".").concat(source));
            }
        },
        getLabel: function (source) {
            // <ArrayInput source="orders" /> => LabelContext is "orders"
            //   <SimpleFormIterator> => LabelContext is ALSO "orders"
            //      <DateInput source="date" /> => use its parent's getLabel so finalLabel = "orders.date"
            //   </SimpleFormIterator>
            // </ArrayInput>
            //
            // we don't prefix with the index to avoid that translation keys contain it
            return parentSourceContext.getLabel(source);
        },
    }); }, [index, parentSourceContext]);
    return (React.createElement(SimpleFormIteratorItemContext_1.SimpleFormIteratorItemContext.Provider, { value: context },
        React.createElement(core_1.SourceContextProvider, { value: sourceContext }, children)));
};
exports.SimpleFormIteratorItemBase = SimpleFormIteratorItemBase;
