// ra-ui-materialui/dist/list/listFieldTypes.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import { DataTable } from "./datatable/index.js";
import { SingleFieldList } from "./SingleFieldList.js";
import { ArrayField, BooleanField, ChipField, DateField, EmailField, ReferenceField, ReferenceArrayField, UrlField, TextField, TextArrayField, } from "../field/index.js";
export var listFieldTypes = {
    table: {
        component: function (props) {
            return React.createElement(DataTable, __assign({}, props));
        },
        representation: function (_props, children) { return "        <DataTable>\n".concat(children.map(function (child) { return "            ".concat(child.getRepresentation()); }).join('\n'), "\n        </DataTable>"); },
    },
    array: {
        component: function (_a) {
            var children = _a.children, props = __rest(_a, ["children"]);
            var childrenArray = React.Children.toArray(children);
            return (React.createElement(DataTable.Col, __assign({}, props),
                React.createElement(ArrayField, __assign({}, props),
                    React.createElement(SingleFieldList, null,
                        React.createElement(ChipField, { source: childrenArray.length > 0 &&
                                React.isValidElement(childrenArray[0]) &&
                                childrenArray[0].props.source })))));
        },
        representation: function (props, children) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <ArrayField source=\"").concat(props.source, "\">\n                    <SingleFieldList>\n                        <ChipField source=\"").concat(children.length > 0 && children[0].getProps().source, "\" />\n                    </SingleFieldList>\n                </ArrayField>\n            </DataTable.Col>");
        },
    },
    scalar_array: {
        component: function (props) {
            return (React.createElement(DataTable.Col, __assign({}, props),
                React.createElement(TextArrayField, __assign({}, props))));
        },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <TextArrayField source=\"").concat(props.source, "\" />\n            </DataTable.Col>");
        },
    },
    boolean: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(BooleanField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <BooleanField source=\"").concat(props.source, "\" />\n            </DataTable.Col>");
        },
    },
    date: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(DateField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <DateField source=\"").concat(props.source, "\" />\n            </DataTable.Col>");
        },
    },
    email: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(EmailField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <EmailField source=\"").concat(props.source, "\" />\n            </DataTable.Col>");
        },
    },
    id: {
        component: function (props) { return React.createElement(DataTable.Col, __assign({}, props)); },
        representation: function (props) { return "<DataTable.Col source=\"".concat(props.source, "\" />"); },
    },
    number: {
        component: DataTable.NumberCol,
        representation: function (props) {
            return "<DataTable.NumberCol source=\"".concat(props.source, "\" />");
        },
    },
    reference: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(ReferenceField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <ReferenceField source=\"").concat(props.source, "\" reference=\"").concat(props.reference, "\" />\n            </DataTable.Col>");
        },
    },
    referenceChild: {
        component: function () { return React.createElement(TextField, { source: "id" }); },
        representation: function () { return "<TextField source=\"id\" />"; },
    },
    referenceArray: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(ReferenceArrayField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <ReferenceArrayField source=\"").concat(props.source, "\" reference=\"").concat(props.reference, "\" />\n            </DataTable.Col>");
        },
    },
    referenceArrayChild: {
        component: function () { return (React.createElement(SingleFieldList, null,
            React.createElement(ChipField, { source: "id" }))); },
        representation: function () {
            return "<SingleFieldList>\n                <ChipField source=\"id\" />\n            </SingleFieldList>";
        },
    },
    richText: undefined, // never display a rich text field in a datagrid
    string: {
        component: DataTable.Col,
        representation: function (props) { return "<DataTable.Col source=\"".concat(props.source, "\" />"); },
    },
    url: {
        component: function (props) { return (React.createElement(DataTable.Col, __assign({}, props),
            React.createElement(UrlField, __assign({}, props)))); },
        representation: function (props) {
            return "<DataTable.Col source=\"".concat(props.source, "\">\n                <UrlField source=\"").concat(props.source, "\" />\n            </DataTable.Col>");
        },
    },
};
